templates/cdk.master.auth-webhook.py

```diff
@@ -0,0 +1,255 @@
+def forward_request(json_req, url):
+    '''Forward a JSON TokenReview request to a url.
+
+    Returns True if the request is authenticated; False if the response is
+    either invalid or authn has been denied.
+    '''
+    try:
+        try:
+            r = requests.post(url, json=json_req)
+        except requests.exceptions.SSLError:
+            app.logger.debug('SSLError with server; skipping cert validation')
+            r = requests.post(url, json=json_req, verify=False)
+    except Exception as e:
+        app.logger.debug('Failed to contact server: {}'.format(e))
+        return False
+
+    # Check if the response is valid
+    try:
+        resp = json.loads(r.text)
+        'authenticated' in resp['status']
+    except (KeyError, TypeError, ValueError):
+        app.logger.debug('Invalid response from server: {}'.format(r.text))
+        return False
+
+    # NB: When forwarding to an external URL, clobber the original request with
+    # the entire server response. This ensures any additional data that the server
+    # wants to send makes it back to the kube apiserver.
+    if resp['status']['authenticated']:
+        json_req = resp
+        return True
+    return False
+
+
+@app.route('/{{ api_ver }}', methods=['POST'])
+def webhook():
+    '''Listen on /$api_version for POST requests.
+
+    For a POSTed TokenReview object, check every known authentication mechanism
+    for a user with a matching token.
+
+    The /$api_version is expected to be the api version of the authentication.k8s.io
+    TokenReview that the k8s-apiserver will be sending.
+
+    Returns:
+        TokenReview object with 'authenticated: True' and user attributes if a
+        token is found; otherwise, a TokenReview object with 'authenticated: False'
+    '''
+    # Log to gunicorn
+    glogger = logging.getLogger('gunicorn.error')
+    app.logger.handlers = glogger.handlers
+    app.logger.setLevel(glogger.level)
+
+    req = request.json
+    try:
+        valid = True if (req['kind'] == 'TokenReview' and
+                         req['spec']['token']) else False
+    except (KeyError, TypeError):
+        valid = False
+
+    if valid:
+        app.logger.debug('REQ: {}'.format(req))
+    else:
+        app.logger.info('Invalid request: {}'.format(req))
+        return ''  # flask needs to return something that isn't None
+
+    # Make the request unauthenticated by deafult
+    req['status'] = {'authenticated': False}
+
+    if (
+        check_token(req)
+        or check_secrets(req)
+        {%- if aws_iam_endpoint %}
+        or check_aws_iam(req)
+        {%- endif %}
+        {%- if keystone_endpoint %}
+        or check_keystone(req)
+        {%- endif %}
+        {%- if custom_authn_endpoint %}
+        or check_custom(req)
+        {%- endif %}
+       ):
+        # Successful checks will set auth and user data in the 'req' dict
+        app.logger.debug('ACK: {}'.format(req))
+    else:
+        app.logger.debug('NAK: {}'.format(req))
```

Comment:
Also need to mask out the password / tokens from the log.  Something like:

```suggestion
from copy import deepcopy
...
    log_req = deepcopy(req)
    log_req['spec']['token'] = '******'
    ...
    if (...):
        app.logger.debug('ACK: {}'.format(log_req))
    else:
        app.logger.debug('NAK: {}'.format(log_req))
```